File: semiprimes.py
# problem 187

limit = 10**8
factors = [0]*limit

factors[1] = 1

def part(n, p):
	count = 0
	while n % p == 0:
		n /= p
		count += 1
	return count

for i in xrange(2, limit):
	if factors[i] == 0:
		for j in xrange(i, limit, i):
			factors[j] += part(j, i)

print len([x for x in factors if x == 2])

File: semiprimes.cpp
// problem 187
#include <iostream>
using namespace std;

#define limit 100000000

int part(int n, int p) {
	int count = 0;
	int q = p;
	while (n % q == 0) {
		q *= p;
		count++;
	}
	return count;
}

static int factors[limit] = {0};

int main(void) {
	factors[1] = 1;
	int count = 0;
	for (int i = 2; i < limit; ++i) {
		if (factors[i] == 0) {
			for (int j = i; j < limit; j += i) {
				factors[j] += part(j, i);
			}
		}
		if (factors[i] == 2) count++;
	}
	cout << count << endl;
	return 0;
}
